### DISCUSSION TITLE
Scenario "Saskatoon": counting IPs.

### DISCUSSION THREAD

**Discussion by fduran** [SadServers/sadservers#18] [Scenarios]
Discussions about the [Saskatoon scenario](https://github.com/fduran/sadservers/tree/main/scenarios/saskatoon)

**Comment by JustPipIt** [comment]
I ran into something interesting when working through this scenario. I issued the following on the file to look for the IP with the most connection requests
`cat access.log | awk '{print $1}' | uniq -c | sort` 
and it showed the IPs based on the number of times they occurred. I piped that result (just the IP) into the highestip.txt file but the solution check said it was wrong.

After getting the solution from the scenario, the correct IP was occurring multiple times but only being seen as occurring 1 time by uniq each time that IP was seen. Whereas the IP my command string was finding was being reported as occurring 97 times in the file. Curious where I may have gone? Thanks in advance.

**Reply by fduran** [reply]
@JustPipIt what IP was your solution? it's possible this is an issue with a newline missing (i wasn't accounting for that in these first scenarios, I should) that is expected like `echo "1.2.3.4" > /home/admin/highestip.txt`

**Reply by JustPipIt** [reply]
The IP was 75.97.9.59

Here are the results in real time
`admin@ip-172-31-27-155:~$ cat access.log  | awk '{print $1}' | uniq -c | sort | tail
     39 130.237.218.86
     41 199.168.96.66
     41 50.139.66.106
     42 130.237.218.86
     49 130.237.218.86
     49 75.97.9.59
     50 75.97.9.59
     85 130.237.218.86
     89 130.237.218.86
     97 75.97.9.59
admin@ip-172-31-27-155:~$ cat access.log  | awk '{print $1}' | uniq -c | sort | tail | awk '{print $2}' | tail -1 > highestip.txt 
admin@ip-172-31-27-155:~$ cat highestip.txt 
75.97.9.59
admin@ip-172-31-27-155:~$ sha1sum highestip.txt 
cf1b80c74c7a7dae55a02d25e6f3ae5402765707  highestip.txt`

**Reply by ivarmedi** [reply]
you're missing a `sort` before the `uniq -c` command:
```
 awk '{print $1}' access.log | sort | uniq -c | sort
```

it's explained in `man uniq`:
>        'uniq' does not detect repeated lines unless they are adjacent.  You may want to sort the input first, or use 'sort -u' without 'uniq'.
